src/TorchSharp/Tensor/Storage.cs

```diff
@@ -0,0 +1,522 @@
+            /// Convert the storage instance to a .NET array, copying its data.
+            /// </summary>
+            /// <returns></returns>
+            public T[] ToArray()
+            {
+                ValidateNotScalar();
+                var result = new T[Count];
+                CopyTo(result, 0);
+                return result;
+            }
+
+            /// <summary>
+            /// Accesses a single element of a storage.
+            /// </summary>
+            /// <param name="index">The index of the element.</param>
+            /// <returns></returns>
+            public T this[int index] {
+                get {
+                    ValidateNotScalar();
+                    CheckIndex(index);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        return ptr[index];
+                    }
+                }
+                set {
+                    ValidateNotScalar();
+                    CheckIndex(index);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        ptr[index] = value;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Accesses a slice of a storage instance.
+            /// </summary>
+            /// <param name="range">The range, expressed as a tuple [start,end)</param>
+            public Storage<T> this[(int? start, int? end) range] {
+                get {
+                    ValidateNotScalar();
+                    var start = CheckIndex(range.start.HasValue? range.start.Value : 0);
+                    var end = CheckIndex(range.end.HasValue ? range.end.Value : Count);
+                    var count = end - start;
+                    unsafe {
+                        var ptr = _tensor_data_ptr + sizeof(T) * start;
+                        return new Storage<T>(ptr, _tensor.dtype, count);
+                    }
+                }
+                set {
+                    ValidateNotScalar();
+                    if (value._tensor is not null || value._tensor_data_ptr != IntPtr.Zero)
+                        throw new InvalidOperationException();
+                    var v = value._scalarValue;
+                    var start = CheckIndex(range.start.HasValue ? range.start.Value : 0);
+                    var end = CheckIndex(range.end.HasValue ? range.end.Value : Count);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        for (int i = start; i < end; i++)
+                            ptr[i] = v;
+                    }
+                }
+            }
+
+#if !NETSTANDARD2_0_OR_GREATER
+            /// <summary>
+            /// Accesses a slice of a storage instance.
+            /// </summary>
+            /// <param name="range">The range.</param>
+            public Storage<T> this[System.Range range] {
+                get {
+                    ValidateNotScalar();
```

Comment:
>ValidateNotScalar [](http://example.com/codeflow?start=20&length=17)

do we need this validation as we already calling the other indexer which already doing the validation?